source/Calamari.Aws/Kubernetes/Discovery/AwsAuthenticationDetails.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Collections.Generic;
+using Amazon.Runtime;
+using Calamari.Common.Features.Discovery;
+
+namespace Calamari.Kubernetes.Aws
+{
+    public class AwsAuthenticationDetails : ITargetDiscoveryAuthenticationDetails
+    {
+        const string DefaultSessionName = "OctopusKubernetesClusterDiscovery";
+        
+        /// <exception cref="AggregateException">
+        /// If both InstanceProfile and EnvironmentVariable Credentials fail.
+        /// Contains AmazonClientExceptions for both InstanceProfile and EnvironmentVariable failures</exception>
+        /// <exception cref="AmazonClientException">If Basic (Account) Credentials fail</exception>
+        public AWSCredentials ToCredentials()
+        {
+            AWSCredentials account;
+            if (Credentials.Type == "account")
+            {
+                account = new BasicAWSCredentials(Credentials.Account.AccessKey, Credentials.Account.SecretKey);
+            }
+            else
+            {
+                try
+                {
+                    // If not currently running on an EC2 instance,
+                    // this will throw an exception.
+                    account = new InstanceProfileAWSCredentials();
+                }
+                catch (Exception instanceProfileException)
+                {
+                    try
+                    {
+                        // The last attempt is trying to use Environment Variables.
+                        account = new EnvironmentVariablesAWSCredentials();
+                    }
+                    catch (Exception environmentVariablesException)
+                    {
+                        throw new AggregateException(instanceProfileException, environmentVariablesException);
```

Comment:
Rather than throwing this general aggregate exception here wrapping two `Exception`s, we should either be throwing a more semantically informative exception.
In this case, I would prefer modifying the return type to use IResult, and return success or failure objects.

Reply:
As discussed, we've updated the method to be a "TryGet" style method and return a bool, it also log it's own exceptions.